// @xan105/is@2.10.1/lib/assert/type/obj.mjs
import { Failure as e } from "../../../../../../../@xan105_error@1.7.1/node_modules/@xan105/error/lib/failure.mjs";
import { asError as i } from "../../opt/obj.mjs";
import { asString as n } from "../../opt/string.mjs";
import { isObjWithinObj as p, isObj as c, isRegExp as f } from "../../is/type/obj.mjs";
function s(t, o = null) {
  if (c(t) === !1)
    throw i(o) ?? new e(n(o) ?? 'Expected "plain obj" ({}, new Object(), Object.create(null))', { code: 1, info: {
      type: typeof t,
      tag: Object.prototype.toString.call(t),
      value: t
    } });
}
function O(t, o = null) {
  if (p(t) === !1)
    throw i(o) ?? new e(n(o) ?? 'Expected "plain obj" ({}, new Object(), Object.create(null)) assigned as property within another', { code: 1, info: {
      type: typeof t,
      tag: Object.prototype.toString.call(t),
      value: t
    } });
}
function g(t, o = null) {
  if (f(t) === !1)
    throw i(o) ?? new e(n(o) ?? "Expected RegExp !", { code: 1, info: {
      type: typeof t,
      tag: Object.prototype.toString.call(t),
      value: t
    } });
}
export {
  s as shouldObj,
  O as shouldObjWithinObj,
  g as shouldRegExp
};
