// @xan105/ffi@1.3.0/lib/koffi/types/windows.mjs
import { arch as y } from "node:process";
import t from "koffi";
const s = y === "x64", I = t.types.void, D = t.types.uint32, o = t.types.long, r = t.types.uint16, P = t.types.int16, i = t.types.uint8, E = t.types.char16, S = t.types.int32, A = t.types.uint16, C = s ? t.pointer(t.types.uint64) : t.pointer(t.types.uint32), n = t.opaque(), p = s ? t.pointer(t.types.int64) : t.pointer(t.types.long), e = s ? t.pointer(t.types.uint64) : t.pointer(t.types.ulong), T = t.types.int, U = t.types.bool, G = t.pointer(t.types.void), W = t.types.uint8, _ = t.types.uint8, l = o, g = t.types.uint64, O = e, B = t.types.uint32, d = t.types.uint64, v = t.types.float, F = n, M = s ? t.types.int32 : P, a = n, K = n, V = n, Y = n, h = n, f = n, Z = n, m = n, b = n, x = n, k = n, q = n, J = n, Q = n, j = n, w = n, z = n, X = n, L = n, $ = n, tt = n, nt = n, st = n, ot = n, pt = L, et = n, it = n, ct = n, yt = t.types.long, rt = t.types.long, Pt = n, Tt = n, Ot = n, Lt = n, Nt = n, Rt = n, ut = t.types.int, Ht = s ? t.types.int64 : t.types.int32, It = t.types.int8, Dt = t.types.int16, Et = t.types.int32, St = t.types.int64, At = r, Ct = o, Ut = o, Gt = o, Wt = t.types.long, N = t.types.longlong, _t = t.types.int32, lt = t.types.int64, gt = p, Bt = T, dt = t.pointer(i), vt = o, Ft = t.types.string, Mt = t.types.string16, at = t.types.string16, c = t.pointer(t.types.void), Kt = c, Vt = t.pointer(t.types.uint16), Yt = s ? t.pointer(t.types.int64) : t.pointer(t.types.int32), ht = t.pointer(t.types.int), ft = t.pointer(t.types.int32), Zt = t.pointer(t.types.void), mt = t.pointer(t.types.void), bt = t.pointer(t.types.char), xt = t.pointer(t.types.uint16), kt = t.pointer(t.types.uint16), qt = t.pointer(t.types.uint16), Jt = p, Qt = t.pointer(t.types.int), jt = t.pointer(t.types.bool), wt = t.pointer(i), zt = t.pointer(t.types.char), Xt = t.pointer(t.types.uint8), $t = s ? t.pointer(t.types.int16) : t.pointer(t.types.int8), tn = t.pointer(t.types.uint16), nn = t.pointer(t.types.uint32), sn = t.pointer(t.types.uint64), on = O, pn = t.pointer(t.types.uint32), en = t.pointer(t.types.uint64), cn = t.pointer(t.types.float), yn = t.pointer(t.types.void), rn = s ? t.pointer(t.pointer(t.types.uint64)) : t.pointer(t.pointer(t.types.uint32)), Pn = s ? t.pointer(t.pointer(t.types.uint64)) : t.pointer(t.pointer(t.types.uint32)), Tn = t.pointer(t.types.int), On = t.pointer(t.pointer(t.types.int)), Ln = t.pointer(t.types.int8), Nn = t.pointer(t.types.int16), Rn = t.pointer(t.types.int32), un = t.pointer(t.types.int64), Hn = t.pointer(t.types.uint32), In = t.pointer(t.types.long), Dn = t.pointer(t.types.int64), En = t.pointer(t.types.void), Sn = t.pointer(t.types.int32), An = t.pointer(t.types.int64), Cn = t.pointer(t.types.int32), Un = s ? t.pointer(t.types.int64) : t.pointer(t.types.int32), Gn = t.pointer(t.types.void), Wn = t.pointer(t.types.void), _n = t.pointer(t.types.int16), ln = e, gn = t.pointer(t.types.void), Bn = t.pointer(t.types.char), dn = t.pointer(t.types.int16), vn = t.pointer(t.types.uint16), Fn = t.pointer(t.types.uint16), Mn = t.pointer(t.types.void), an = t.pointer(t.types.void), Kn = t.pointer(t.types.uint), Vn = t.pointer(t.pointer(t.types.uint)), Yn = t.pointer(t.types.uint8), hn = t.pointer(t.types.uint16), fn = t.pointer(t.types.uint32), Zn = t.pointer(t.types.uint64), mn = t.pointer(t.types.uint), bn = t.pointer(t.types.uint64), xn = t.pointer(t.pointer(t.types.uint64)), kn = t.pointer(t.types.uint), qn = t.pointer(t.types.uint64), Jn = t.pointer(t.types.uint16), Qn = t.pointer(t.types.uint16), jn = t.pointer(t.types.uint16), wn = t.pointer(t.types.uint16), zn = t.pointer(t.types.uint64), Xn = n, $n = c, ts = n, ns = e, ss = p, os = t.types.int16, ps = t.types.uint16, es = t.types.uchar, is = s ? t.types.uint32 : t.types.uint16, cs = t.types.uint, R = s ? t.types.uint64 : t.types.uint32, ys = t.types.uint8, rs = t.types.uint16, Ps = t.types.uint32, Ts = t.types.uint64, Os = t.types.ulong, Ls = t.types.ulonglong, Ns = t.types.uint32, Rs = t.types.uint64, us = t.types.ushort, Hs = N, Is = R;
export {
  S as ACCESS_MASK,
  A as ATOM,
  T as BOOL,
  U as BOOLEAN,
  i as BYTE,
  G as CALLBACK,
  W as CCHAR,
  _ as CHAR,
  l as COLORREF,
  o as DWORD,
  B as DWORD32,
  d as DWORD64,
  g as DWORDLONG,
  O as DWORD_PTR,
  D as ENUM,
  v as FLOAT,
  F as HACCEL,
  M as HALF_PTR,
  n as HANDLE,
  a as HBITMAP,
  K as HBRUSH,
  V as HCOLORSPACE,
  Y as HCONV,
  h as HCONVLIST,
  f as HCURSOR,
  Z as HDC,
  m as HDDEDATA,
  b as HDESK,
  x as HDROP,
  k as HDWP,
  q as HENHMETAFILE,
  J as HFILE,
  Q as HFONT,
  j as HGDIOBJ,
  w as HGLOBAL,
  z as HHOOK,
  X as HICON,
  L as HINSTANCE,
  $ as HKEY,
  tt as HKL,
  nt as HLOCAL,
  st as HMENU,
  ot as HMETAFILE,
  pt as HMODULE,
  et as HMONITOR,
  it as HPALETTE,
  ct as HPEN,
  yt as HRESULT,
  Pt as HRGN,
  Tt as HRSRC,
  Ot as HSZ,
  Lt as HWINEVENTHOOK,
  Nt as HWINSTA,
  Rt as HWND,
  ut as INT,
  Dt as INT16,
  Et as INT32,
  St as INT64,
  It as INT8,
  Ht as INT_PTR,
  At as LANGID,
  Ct as LCID,
  Ut as LCTYPE,
  Gt as LGRPID,
  Wt as LONG,
  _t as LONG32,
  lt as LONG64,
  N as LONGLONG,
  p as LONG_PTR,
  gt as LPARAM,
  Bt as LPBOOL,
  dt as LPBYTE,
  vt as LPCOLORREF,
  Ft as LPCSTR,
  Mt as LPCTSTR,
  Kt as LPCVOID,
  at as LPCWSTR,
  Vt as LPDWORD,
  Yt as LPHANDLE,
  ht as LPINT,
  ft as LPLONG,
  Zt as LPMSG,
  mt as LPPOINT,
  bt as LPSTR,
  kt as LPTSTR,
  c as LPVOID,
  qt as LPWORD,
  xt as LPWSTR,
  Jt as LRESULT,
  rt as NTSTATUS,
  Qt as PBOOL,
  jt as PBOOLEAN,
  wt as PBYTE,
  zt as PCHAR,
  Xt as PCSTR,
  $t as PCTSTR,
  tn as PCWSTR,
  nn as PDWORD,
  pn as PDWORD32,
  en as PDWORD64,
  sn as PDWORDLONG,
  on as PDWORD_PTR,
  cn as PFLOAT,
  yn as PHALF_PTR,
  rn as PHANDLE,
  Pn as PHKEY,
  Tn as PINT,
  Nn as PINT16,
  Rn as PINT32,
  un as PINT64,
  Ln as PINT8,
  On as PINT_PTR,
  Hn as PLCID,
  In as PLONG,
  Sn as PLONG32,
  An as PLONG64,
  Dn as PLONGLONG,
  En as PLONG_PTR,
  Cn as POINTER_32,
  Un as POINTER_64,
  Gn as POINTER_SIGNED,
  Wn as POINTER_UNSIGNED,
  _n as PSHORT,
  ln as PSIZE,
  gn as PSSIZE,
  Bn as PSTR,
  dn as PTBYTE,
  vn as PTCHAR,
  Fn as PTSTR,
  Mn as PUCHAR,
  an as PUHALF_PTR,
  Kn as PUINT,
  hn as PUINT16,
  fn as PUINT32,
  Zn as PUINT64,
  Yn as PUINT8,
  Vn as PUINT_PTR,
  mn as PULONG,
  kn as PULONG32,
  qn as PULONG64,
  bn as PULONGLONG,
  xn as PULONG_PTR,
  Jn as PUSHORT,
  C as PVOID,
  Qn as PWCHAR,
  jn as PWORD,
  wn as PWSTR,
  zn as QWORD,
  Xn as SC_HANDLE,
  $n as SC_LOCK,
  ts as SERVICE_STATUS_HANDLE,
  P as SHORT,
  ns as SIZE,
  ss as SSIZE,
  os as TBYTE,
  ps as TCHAR,
  es as UCHAR,
  is as UHALF_PTR,
  cs as UINT,
  rs as UINT16,
  Ps as UINT32,
  Ts as UINT64,
  ys as UINT8,
  R as UINT_PTR,
  Os as ULONG,
  Ns as ULONG32,
  Rs as ULONG64,
  Ls as ULONGLONG,
  e as ULONG_PTR,
  us as USHORT,
  Hs as USN,
  I as VOID,
  E as WCHAR,
  r as WORD,
  Is as WPARAM
};
