// @xan105/error@1.7.1/lib/failure.mjs
import { filterStackTrace as s } from "./util/clean.mjs";
import { CODES as i } from "./codes/custom.mjs";
class n extends Error {
  constructor(c, e = {}) {
    (typeof e == "string" || Number.isSafeInteger(e)) && (e = { code: e });
    const r = {
      code: e.code ?? null,
      cause: e.cause || null,
      clean: e.clean ?? !0,
      filter: Array.isArray(e.filter) && e.filter.every((t) => typeof t == "string" && t.length > 0) ? e.filter : [],
      info: typeof e.info == "object" || typeof e.info == "string" ? e.info : null
      //object as in object,array,etc... Or just a string
    };
    r.cause instanceof Error ? super(c || "An error has occurred", { cause: r.cause }) : super(c || "An error has occurred"), typeof r.code == "string" && r.code.length > 0 ? this.code = r.code : Number.isSafeInteger(r.code) && (this.code = i[r.code] ?? `Error ${r.code}`), this.name = this.constructor.name, this.code && (this.name += " [" + this.code + "]"), r.info && (this.info = r.info), Error.captureStackTrace && (Error.captureStackTrace(this, this.constructor), r.clean && (this.stack = s(this.stack, r.filter)));
  }
}
export {
  n as Failure
};
