// @xan105/ffi@1.3.0/lib/koffi/open.mjs
import { platform as k } from "node:process";
import { promisify as w } from "node:util";
import _ from "koffi";
import { Failure as y } from "../../../../../../@xan105_error@1.7.1/node_modules/@xan105/error/lib/failure.mjs";
import { attemptify as g } from "../../../../../../@xan105_error@1.7.1/node_modules/@xan105/error/lib/attempt.mjs";
import { asArray as p } from "../../../../../../@xan105_is@2.10.1/node_modules/@xan105/is/lib/opt/array.mjs";
import { shouldObjWithinObj as F, shouldObj as h } from "../../../../../../@xan105_is@2.10.1/node_modules/@xan105/is/lib/assert/type/obj.mjs";
import { asStringLike as S } from "../../../../../../@xan105_is@2.10.1/node_modules/@xan105/is/lib/opt/string.mjs";
import { asBoolean as o } from "../../../../../../@xan105_is@2.10.1/node_modules/@xan105/is/lib/opt/boolean.mjs";
import { shouldStringNotEmpty as j } from "../../../../../../@xan105_is@2.10.1/node_modules/@xan105/is/lib/assert/type/string.mjs";
import "../../../../../../@xan105_is@2.10.1/node_modules/@xan105/is/lib/is/file/binary.mjs";
import { conventions as A } from "./util/abi.mjs";
import { hashFileSync as I } from "../util.mjs";
function O(i, n = {}) {
  j(i), h(n);
  const e = {
    ignoreLoadingFail: o(n.ignoreLoadingFail) ?? !1,
    ignoreMissingSymbol: o(n.ignoreMissingSymbol) ?? !1,
    lazy: o(n.lazy) ?? !1,
    global: o(n.global) ?? !1,
    abi: A.includes(n.abi) ? n.abi : "cdecl",
    integrity: S(n.integrity, "SRI") ?? ""
  }, l = {
    win32: ".dll",
    darwin: ".dylib"
  }[k] ?? ".so";
  if (i.endsWith(l) || (i += l), e.integrity) {
    const [r, c] = e.integrity.split("-"), [a = ""] = g(I)(i, r);
    if (a !== c)
      throw new y("Integrity check failed ! Dylib loading has been aborted.", {
        code: "ERR_INTEGRITY",
        info: { path: i, algo: r, hash: a }
      });
  }
  const [t, d] = g(_.load)(i, {
    lazy: e.lazy,
    global: e.global
  });
  return function(r, c, a) {
    try {
      if (d) throw d;
      return t.func("__" + e.abi, r, c, a);
    } catch (f) {
      if (e.ignoreMissingSymbol && f.message.startsWith("Cannot find function") && f.message.endsWith("in shared library") || !t && e.ignoreLoadingFail) return;
      throw new y(f.message, {
        code: "ERR_FFI",
        cause: f,
        info: { lib: i, symbol: r }
      });
    }
  };
}
function G(i, n, e = {}) {
  F(n), h(e);
  const l = {
    errorAtRuntime: o(e.errorAtRuntime) ?? !1,
    nonblocking: o(e.nonblocking) ?? !1,
    stub: o(e.stub) ?? !1
  };
  delete e.errorAtRuntime, delete e.nonblocking, delete e.stub;
  const t = /* @__PURE__ */ Object.create(null), d = O(i, e);
  for (const [s, r] of Object.entries(n)) {
    if (s === "__proto__") continue;
    const c = r.symbol || s, a = r.result || "void", f = p(r.parameters) ?? [], b = o(r.nonblocking) ?? l.nonblocking, R = o(r.stub) ?? l.stub, [u, m] = g(d)(c, a, f);
    if (m && l.errorAtRuntime === !1) throw m;
    typeof u == "function" ? (t[s] = b ? w(u.async) : u, delete t[s].async) : m ? t[s] = b ? () => Promise.reject(m) : () => {
      throw m;
    } : R && (t[s] = b ? () => Promise.resolve() : () => {
    });
  }
  return t;
}
export {
  G as dlopen,
  O as load
};
